tests/common.rs

```diff
@@ -0,0 +1,197 @@
+    /// Get the address of the nats instance
+    async fn get_addr(&self) -> Result<String>;
+}
+
+const NUM_PORT_TRIES: u8 = 5;
+
+/// Helper function that picks a random port
+fn pick_random_port() -> Result<u16> {
+    let mut tries = 0;
+    let mut port = None;
+    loop {
+        tries += 1;
+        if let Some(p) = random_free_tcp_port() {
+            port = Some(p);
+            break;
+        }
+        if tries >= 5 {
+            return Err(Error::msg(format!("Failed to find port after trying {NUM_PORT_TRIES} times")));
+        }
+    }
+
+    match port {
+        Some(port) => Ok(port),
+        None => Err(Error::msg("Failed to find free TCP port")),
+    }
+}
+
+/// Start a NATS instance for use during test
+pub(crate) async fn make_nats_instance() -> Result<Box<dyn NatsInstance>> {
+    let in_ci = env::var(ENV_VAR_CI).is_ok();
+    let docker_disabled = env::var(ENV_VAR_DOCKER_DISABLED).map(|v| v == "true").unwrap_or(DEFAULT_DOCKER_DISABLED);
+
+    let port = pick_random_port()?;
+
+    if docker_disabled || in_ci {
+        Ok(Box::new(ProcessNats::new(port).await?))
+    } else {
+        Ok(Box::new(DockerizedNats::new(port).await?))
+    }
+}
+
+/// A process-based NATS instance
+struct ProcessNats {
+    child: Child,
+    port: u16,
+}
+
+impl ProcessNats {
+    async fn new(port: u16) -> Result<ProcessNats> {
+        // Determine path to NATS binary
+        let nats_bin_path = env::var(ENV_VAR_NATS_BIN_PATH).unwrap_or(DEFAULT_NATS_BIN_PATH.into());
+
+        // Start process in a local process
+        let child = Command::new(&nats_bin_path)
+            .arg("-js")
+            .arg("-p")
+            .arg(port.to_string())
+            .stdout(std::process::Stdio::null())
+            .stderr(std::process::Stdio::null())
+            .spawn()
+            .map_err(|e| Error::msg(format!("nats command ({nats_bin_path}) failed to spawn: {e}")))?;
+
+        // Wait until NATS is available (usually ~1s)
+        tokio::time::sleep(tokio::time::Duration::from_millis(1000)).await;
+
+        Ok(ProcessNats { child, port })
+    }
+}
+
+#[async_trait]
+impl EphemeralResource for ProcessNats {
+    /// Stop the instance
+    async fn stop(&mut self) -> Result<()> {
+        self.child.kill()
+            .await
+            .map_err(|e| Error::msg(format!("failed to stop NATS child process: {e}")))
+    }
+}
+
+#[async_trait]
+impl NatsInstance for ProcessNats {
+    async fn get_addr(&self) -> Result<String> {
+        Ok(format!("localhost:{}", self.port))
+    }
+}
+
+/// A dockerized NATS instance
+struct DockerizedNats {
+    /// Docker instance
+    docker: Docker,
+
+    /// Port on which to start the NATS container
+    port: u16,
+
+    /// Docker container ID
+    id: String,
+
+    // Container name
+    container_name: String
+}
+
+const CONTAINER_SUFFIX_ALPHABET: &str = "01234567890";
+
+impl DockerizedNats {
+    async fn new(port: u16) -> Result<DockerizedNats> {
```

Comment:
see my previous comment. It feels like we're incurring a pile of overhead in launching docker and NATS etc when environment scripts like Makefiles, bash scripts, and github workflows should be doing this

Reply:
The thinking here was that using the methods up til now require running serially -- since the NATS is shared between tests they have to run serially. In general the only way past this is running setup per-test or per-suite.

The reason I needed two ways to do this was because Github CI doesn't have docker dind (otherwise it could be consistently docker across local and CI environments and a lot more "contained") -- also the alternative was nice for people who don't like using `docker`.

Reply:
When I need to run stuff serially, I just make sure that I run cargo test with a single test thread.